gems/smithy-client/spec/smithy-client/waiters_spec.rb

```diff
@@ -0,0 +1,1364 @@
+                    'key3' => 'abc'
+                  }
+                }
+                client.stub_responses(:get_widget, output)
+                expect do
+                  client.wait_until(:projection_matcher, input, max_wait_time: 60)
+                end.to_not raise_error
+              end
+
+              it 'fails when not matched' do
+                output = {
+                  data_map: {
+                    'key1' => 'abc',
+                    'key2' => 'def',
+                    'key3' => 'ghi'
+                  }
+                }
+                client.stub_responses(:get_widget, output)
+                expect do
+                  client.wait_until(:projection_matcher, input, max_wait_time: 0)
+                end.to raise_error(max_wait_time_exceeded_error)
+              end
+            end
+
+            context 'contains field' do
+              before(:each) do
+                shapes['smithy.ruby.tests#GetWidget']['traits']['smithy.waiters#waitable'] = {
+                  'ContainsFieldMatcher' => {
+                    'acceptors' => [
+                      {
+                        'state' => 'success',
+                        'matcher' => {
+                          'output' => {
+                            'path' => 'contains(dataMap.*, stringProperty)',
+                            'expected' => 'true',
+                            'comparator' => 'booleanEquals'
+                          }
+                        }
+                      }
+                    ]
+                  }
+                }
+              end
+
+              it 'succeeds when matched' do
+                output = {
+                  string_property: 'match',
+                  data_map: {
+                    'key1' => 'not a match',
+                    'key2' => 'match',
+                    'key3' => 'not a match'
+                  }
+                }
+                client.stub_responses(:get_widget, output)
+                expect do
+                  client.wait_until(:contains_field_matcher, input, max_wait_time: 60)
+                end.to_not raise_error
+              end
+
+              it 'fails when not matched' do
+                output = {
+                  string_property: 'match',
+                  data_map: {
+                    'key1' => 'not a match',
+                    'key2' => 'not a match',
+                    'key3' => 'not a match'
+                  }
+                }
+                client.stub_responses(:get_widget, output)
+                expect do
+                  client.wait_until(:contains_field_matcher, input, max_wait_time: 0)
+                end.to raise_error(max_wait_time_exceeded_error)
+              end
+            end
+
+            context 'and inequality' do
+              before(:each) do
+                shapes['smithy.ruby.tests#GetWidget']['traits']['smithy.waiters#waitable'] = {
+                  'AndInequalityMatcher' => {
```

Comment:
This is a fundamental difference I think between kotlin's approach and ruby - I don't know if we care to test things like flatten/and inequality, etc - These are really testing JMESPath implementations. I don't think Kotlin has a runtime JMESPath implementation like we do. I think we care to test the basic waiter matcher types (like booleanEquals) and we can rely on the fact that JMESPath handles "&&" expressions without having to test it here. So I think you can cut a good chunk of the tests in this file.